### DISCUSSION TITLE
Put arguments to new lines for long function signatures

### DISCUSSION THREAD

**Discussion by tuukkamustonen** [pradyunsg/furo#388] [Ideas] [answered]
With long function signatures you get something like:

![image](https://user-images.githubusercontent.com/94327/156140424-48eb5993-718b-4af7-a315-ac3b4949d4ca.png)

It's quite ugly to look at, better would be:

```py
class pastry.auth.token_provider.OAuth2TokenProvider(
    url: str,
    client_id: Optional[str] = None,
    client_secret: Optional[str] = None,
    ...and so on...
)
```

There's a ticket also for Sphinx about the same thing at https://github.com/sphinx-doc/sphinx/issues/1514 but this is perhaps a representation issue, rather, and should be discussed per theme.

Hacking the CSS I could probably make each parameter show on a new line, but that would then apply also for functions with only 1-2 parameters, too.

Would detecting long signatures be difficult?

**Comment by pradyunsg** [comment]
This isn’t feasible in CSS or the theme side without invasive hacks. It has to happen on Sphinx’s end.

**Reply by tuukkamustonen** [reply]
Impressive response time 🤗

I think even then "detecting a long line" might be problematic as styling then makes the lines various lengths. Alright.

**Reply by cjw296** [reply]
https://github.com/sphinx-doc/sphinx/issues/1514#issuecomment-740770081 seems to imply the Sphinx side is done @pradyunsg?

**Reply by wasi-master** [reply]
I could get it somewhat working with the custom CSS shown in [this comment](https://github.com/sphinx-doc/sphinx/issues/1514#issuecomment-742703082)

![image](https://user-images.githubusercontent.com/63045920/168410837-ced13fa7-e0b5-4249-bff8-439090b2faa0.png)

to

![image](https://user-images.githubusercontent.com/63045920/168410822-4e62c8c4-fa61-4dc4-aecb-ec756797662a.png)

**Comment by cjw296** [comment]
@pradyunsg - I tagged you on https://github.com/sphinx-doc/sphinx/issues/1514#issuecomment-1175868713, which I think is the closest I've found to a relevant open issue for Sphinx itself.

I guess we really want this to be intelligent, no point in putting short argument definitions on multiple lines, but really want to see long signatures split. Doing whatever is done in the source might be a good option...

Anyway, I have a [particularly ugly example](https://testfixtures.readthedocs.io/en/latest/api.html#testfixtures.compare) now, so keen to help where I can or nudge things along where I can't ;-)

**Reply by pradyunsg** [reply]
FWIW, @cjw296 I just used that "ugly" example, to advocate for the behaviours in a PR about this. ;)

https://github.com/sphinx-doc/sphinx/pull/11011#discussion_r1061036633

**Comment by pradyunsg** [comment, marked as answer]
https://github.com/sphinx-doc/sphinx/pull/11011 seems like the way forward here. Once that is implemented, I'd be happy to make sure that Furo handles that output gracefully. :)

**Reply by cjw296** [reply]
@pradyunsg - cool, should we open an issue in this repo to track? Lots of work seems to be happening on that PR so 🤞

**Reply by pradyunsg** [reply]
No. It should "just work" last I checked, and if it doesn't, file a discussion as a potential issue for that.

**Reply by cjw296** [reply]
🤞
